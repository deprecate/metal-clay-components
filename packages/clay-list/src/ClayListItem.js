import 'clay-checkbox';
import 'clay-icon';
import 'clay-label';
import 'clay-link';
import 'clay-sticker';
// eslint-disable-next-line
import { ClayActionsDropdown } from 'clay-dropdown';
import {Config} from 'metal-state';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayListItem.soy.js';

/**
 * Metal ClayListItem component.
 */
class ClayListItem extends Component {
	/**
	 * Continues the propagation of the checkbox changed event
	 * @param {!Event} event
	 * @private
	 */
	handleItemCheckboxClick_(event) {
		this.emit('itemToggled', event);
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayListItem.STATE = {
	/**
	 * List of items to display in the actions menu.
	 * @instance
	 * @memberof ClayListItem
	 * @type {!array}
	 * @default undefined
	 */
	actionItems: Config.arrayOf(
		Config.shapeOf({
			disabled: Config.bool().value(false),
			href: Config.string().required(),
			icon: Config.string(),
			label: Config.string().required(),
			quickAction: Config.bool(),
			separator: Config.bool().value(false),
		})
	),

	/**
	 * Name of the content renderer to use template variants.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	contentRenderer: Config.string(),

	/**
	 * Description of the item
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	description: Config.string(),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Url of the item.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	href: Config.string(),

	/**
	 * Icon of the list item.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	icon: Config.string(),

	/**
	 * Shape of the icon of the list item.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default rounded
	 */
	iconShape: Config.oneOf(['circle', 'rounded']).value('rounded'),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Label of the list item.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	label: Config.string(),

	/**
	 * Label styles map.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?object|undefined}
	 * @default undefined
	 */
	labelStylesMap: Config.object(),

	/**
	 * Flag to indicate if the list item is selectable.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?bool|undefined}
	 * @default false
	 */
	selectable: Config.bool().value(false),

	/**
	 * Flag to indicate if the list item is selected.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?bool|undefined}
	 * @default false
	 */
	selected: Config.bool().value(false),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	spritemap: Config.string(),

	/**
	 * Title of the list item.
	 * @instance
	 * @memberof ClayListItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	title: Config.string().required(),
};

defineWebComponent('clay-list-item', ClayListItem);

Soy.register(ClayListItem, templates);

export {ClayListItem};
export default ClayListItem;
