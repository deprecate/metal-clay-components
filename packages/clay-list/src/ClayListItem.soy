{namespace ClayListItem}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param title: string}
	{@param? actionItems: list<[
			disabled: bool,
			href: string,
			icon: string,
			label: string,
			quickAction: bool,
			separator: bool
	]>}
	{@param? contentRenderer: string}
	{@param? description: string}
	{@param? elementClasses: string}
	{@param? handleItemCheckboxClick_: any}
	{@param? href: string}
	{@param? icon: string}
	{@param? iconShape: string}
	{@param? id: string}
	{@param? label: string}
	{@param? labelStylesMap: ?}
	{@param? selectable: bool}
	{@param? selected: bool}
	{@param? spritemap: string}

	{let $listItemAttributes kind="attributes"}
    class="list-group-item list-group-item-flex
  		{if $elementClasses}
	  		{sp}{$elementClasses}
		  {/if}

  		{if $selected}
	  		{sp}active
		  {/if}
    "

    {if $id}
      id="{$id}"
    {/if}
	{/let}

	<li {$listItemAttributes}>
		{let $isSimple: not $description and not $selectable and not $icon and not $actionItems and not $label/}

		{if $isSimple}
			{delcall ClayListItem.Simple variant="$contentRenderer"}
				{param href: $href /}
				{param title: $title /}
			{/delcall}
		{else}
			{if $selectable}
				<div class="flex-col">
					{call ClayCheckbox.render}
						{param checked: $selected /}
						{param events: ['change': $handleItemCheckboxClick_] /}
						{param hideLabel: true /}
						{param label: 'select' /}
					{/call}
				</div>
			{/if}

			{if $icon and $spritemap}
				{delcall ClayListItem.Icon variant="$contentRenderer"}
					{param icon: $icon /}
					{param iconShape: $iconShape /}
					{param spritemap: $spritemap /}
				{/delcall}
			{/if}

			{delcall ClayListItem.Content variant="$contentRenderer"}
				{param description: $description /}
				{param href: $href /}
				{param label: $label /}
				{param labelStylesMap: $labelStylesMap /}
				{param spritemap: $spritemap /}
				{param title: $title /}
			{/delcall}

			{if $actionItems and $spritemap}
        <div class="flex-col">
					{call .quickMenu}
						{param items: $actionItems /}
						{param spritemap: $spritemap /}
					{/call}

					{call ClayActionsDropdown.render}
						{param items: $actionItems /}
						{param spritemap: $spritemap /}
					{/call}
				</div>
			{/if}
		{/if}
	</li>
{/template}

/**
 * This renders the quick actions menu
 */
{template .quickMenu}
	{@param items: list<[
		disabled: bool,
		href: string,
		icon: string,
		label: string,
		quickAction: bool,
		separator: bool
	]>}
	{@param spritemap: string}

	<div class="quick-action-menu">
		{foreach $item in $items}
			{if $item.quickAction and $item.icon and $spritemap}
				{call ClayLink.render}
					{param elementClasses: 'quick-action-item' /}
					{param href: $item.href /}
					{param icon: $item.icon /}
					{param spritemap: $spritemap /}
				{/call}
			{/if}
		{/foreach}
	</div>
{/template}

/**
 * This renders the content of the item.
 */
{deltemplate ClayListItem.Content}
	{@param title: string}
	{@param? description: string}
	{@param? href: string}
	{@param? label: string}
	{@param? labelStylesMap: ?}
	{@param? spritemap: string}

	<div class="flex-col flex-col-expand">
		<h4 class="list-group-title text-truncate">
			{if $href}
				{call ClayLink.render}
					{param href: $href /}
					{param label: $title /}
				{/call}
			{else}
				{$title}
			{/if}
		</h4>

		{if $description}
			<p class="list-group-subtitle text-truncate">{$description}</p>
		{/if}

		{if $label}
			<div class="list-group-detail">
				{call ClayLabel.render}
					{param label: $label /}
					{param style: $labelStylesMap ? $labelStylesMap[$label] ?: $labelStylesMap['*'] : null /}
				{/call}
			</div>
		{/if}
	</div>
{/deltemplate}

/**
 * This renders the icon of the item.
 */
{deltemplate ClayListItem.Icon}
	{@param icon: string}
	{@param spritemap: string}
	{@param? iconShape: string}

	<div class="flex-col">
		{call ClaySticker.render}
			{param icon: $icon /}
			{param shape: $iconShape /}
			{param spritemap: $spritemap /}
		{/call}
	</div>
{/deltemplate}

/**
 * This renders a simple item.
 */
{deltemplate ClayListItem.Simple}
	{@param title: string}
	{@param? href: string}

	{if $href}
		{call ClayLink.render}
			{param href: $href /}
			{param label: $title /}
		{/call}
	{else}
		{$title}
	{/if}
{/deltemplate}