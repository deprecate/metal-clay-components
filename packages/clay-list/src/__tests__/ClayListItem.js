import ClayListItem from '../ClayListItem';

const spritemap = '../node_modules/lexicon-ux/build/images/icons/icons.svg';

let component;

let actionItems = [
	{
		href: '#1',
		label: 'Option 1',
	},
	{
		href: '#2',
		label: 'Option 2',
		separator: true,
	},
	{
		href: '#3',
		label: 'Option 3',
	},
];

let actionItemsWithQuickItems = [
	{
		href: '#1',
		icon: 'trash',
		label: 'Option 1',
		quickAction: true,
	},
	{
		href: '#2',
		icon: 'download',
		label: 'Option 2',
		quickAction: true,
		separator: true,
	},
	{
		href: '#3',
		icon: 'info-circle-open',
		label: 'Option 3',
		quickAction: true,
	},
];

describe('ClayListItem', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayListItem({
			title: 'My Item',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with classes', () => {
		component = new ClayListItem({
			elementClasses: 'my-custom-class',
			title: 'My Item',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with id', () => {
		component = new ClayListItem({
			id: 'myId',
			title: 'My Item',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with link', () => {
		component = new ClayListItem({
			href: '#1',
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with description', () => {
		component = new ClayListItem({
			description: 'Description 1',
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with description and url', () => {
		component = new ClayListItem({
			description: 'Description 1',
			href: '#1',
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayListItem', () => {
		component = new ClayListItem({
			description: 'Description 1',
			href: '#1',
			selectable: true,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selected ClayListItem', () => {
		component = new ClayListItem({
			description: 'Description 1',
			href: '#1',
			selectable: true,
			selected: true,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with icon', () => {
		component = new ClayListItem({
			icon: 'folder',
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with circle icons', () => {
		component = new ClayListItem({
			icon: 'folder',
			iconShape: 'circle',
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with label', () => {
		component = new ClayListItem({
			label: 'Status',
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with labels with style', () => {
		component = new ClayListItem({
			label: 'Status',
			labelStylesMap: {
				'*': 'info',
			},
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with action menu', () => {
		component = new ClayListItem({
			actionItems: actionItems,
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem with quick action menu', () => {
		component = new ClayListItem({
			actionItems: actionItemsWithQuickItems,
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListItem and emit an event on item toggled', () => {
		const spy = jest.fn();

		component = new ClayListItem({
			events: {itemToggled: spy},
			selectable: true,
			title: 'Item 1',
		});

		component.element.querySelector('input[type="checkbox"]').click();

		expect(spy).toHaveBeenCalled();
	});

	it('should fail when no title is passed', function() {
		expect(() => {
			component = new ClayListItem();
		}).toThrow();
	});
});
