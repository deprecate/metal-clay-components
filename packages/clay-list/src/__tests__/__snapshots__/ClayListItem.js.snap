// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayListItem should render a ClayListItem with action menu 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
  <div class="flex-col">
    <div class="quick-action-menu"></div>
    <div class="dropdown dropdown-action">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
          <title>ellipsis-v</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#ellipsis-v"></use>
        </svg>
      </button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class=" dropdown-item" href="#1">Option 1</a>
          </li>
          <li>
            <a class=" dropdown-item" href="#2">Option 2</a>
          </li>
          <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
          <li>
            <a class=" dropdown-item" href="#3">Option 3</a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with circle icons 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col">
    <span class="sticker sticker-primary rounded-circle">
      <span class="sticker-overlay">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <title>folder</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#folder"></use>
        </svg>
      </span>
    </span>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with classes 1`] = `<li class="list-group-item list-group-item-flex my-custom-class">My Item</li>`;

exports[`ClayListItem should render a ClayListItem with description 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with description and url 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">
      <a href="#1">Item 1</a>
    </h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with icon 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col">
    <span class="sticker sticker-primary">
      <span class="sticker-overlay">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <title>folder</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#folder"></use>
        </svg>
      </span>
    </span>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with id 1`] = `<li class="list-group-item list-group-item-flex" id="myId">My Item</li>`;

exports[`ClayListItem should render a ClayListItem with label 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
    <div class="list-group-detail">
      <span class="label label-secondary">Status</span>
    </div>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with labels with style 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
    <div class="list-group-detail">
      <span class="label label-info">Status</span>
    </div>
  </div>
</li>
`;

exports[`ClayListItem should render a ClayListItem with link 1`] = `
<li class="list-group-item list-group-item-flex">
  <a href="#1">Item 1</a>
</li>
`;

exports[`ClayListItem should render a ClayListItem with quick action menu 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
  <div class="flex-col">
    <div class="quick-action-menu">
      <a class=" quick-action-item" href="#1">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash">
          <title>trash</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#trash"></use>
        </svg>
      </a>
      <a class=" quick-action-item" href="#2">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-download">
          <title>download</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#download"></use>
        </svg>
      </a>
      <a class=" quick-action-item" href="#3">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle-open">
          <title>info-circle-open</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle-open"></use>
        </svg>
      </a>
    </div>
    <div class="dropdown dropdown-action">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
          <title>ellipsis-v</title>
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#ellipsis-v"></use>
        </svg>
      </button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class=" dropdown-item" href="#1">Option 1</a>
          </li>
          <li>
            <a class=" dropdown-item" href="#2">Option 2</a>
          </li>
          <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
          <li>
            <a class=" dropdown-item" href="#3">Option 3</a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</li>
`;

exports[`ClayListItem should render a selectable ClayListItem 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col">
    <div class="custom-control custom-checkbox">
      <label>
        <input class="custom-control-input" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description sr-only">select</span>
      </label>
    </div>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">
      <a href="#1">Item 1</a>
    </h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListItem should render a selected ClayListItem 1`] = `
<li class="list-group-item list-group-item-flex active">
  <div class="flex-col">
    <div class="custom-control custom-checkbox">
      <label>
        <input checked="" class="custom-control-input" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description sr-only">select</span>
      </label>
    </div>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">
      <a href="#1">Item 1</a>
    </h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListItem should render the default markup 1`] = `<li class="list-group-item list-group-item-flex">My Item</li>`;
